Rover steering wheel

noether opened the pull request:
This PR adds support to rovers with two actuators; namely, thrust (both directions) and steering wheel.

The four major (independent) points are:
- New firmware for the vehicle based on the existing firmware for rovers.
- We have added a NPS simulator for the rover. It behaves very similarly to a rover traveling at around 20km/h on a flat world.
- We have modified the "classic" GVF. Now you can call it from different firmware (fixed-wing and rover steering wheel).
- We have added UCM directory in flightplans and airframefiles for our configurations.

noether commented:
Oh, I see that we forgot to implement the speed controller. We will upload it after a couple of tests. Then, we will move the current "Assisted mode" to "Nav mode". The assisted mode will disappear probably and we will have only two modes, manual and auto2/nav.

Anyway, there is a lot of code to check by you guys :P.

gautierhattenberger commented on conf/messages.xml:
```diff
@@ -0,0 +1,3221 @@
+<?xml version="1.0"?>
+<!DOCTYPE protocol SYSTEM "messages.dtd">
```
remove this file from conf folder
if you need a new message, make a pull request in pprzlink and update the reference in paparazzi submodule

gautierhattenberger commented on conf/modules/board_matek_f765_car.xml:
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+
+<module name="board_matek_f765_car" dir="boards">
```
what is the difference with "wing" version ?

jesusBV20 replied:
There are no differences. I had to remove that file

gautierhattenberger commented on conf/modules/fdm_rover.xml:
```diff
@@ -0,0 +1,14 @@
+<module name="fdm_rover" dir="fdm">
+  <doc>
+    <description>
+
```
please add a short description

gautierhattenberger commented on conf/radios/UCM/T16SZ_SBUS_rover.xml:
```diff
@@ -0,0 +1,64 @@
+
+<!DOCTYPE radio SYSTEM "../radio.dtd">
+<radio name="Futaba T16SZ with SBUS" data_min="1000" data_max="2000" sync_min ="5000" sync_max ="15000" pulse_type="POSITIVE">
+ <channel ctl="A" function="ROLL"     min="1000" neutral="1500" max="2000" average="0"/>
```
the `ctl` attribute is no longer needed and can be removed

gautierhattenberger commented on conf/telemetry/steering_rover.xml:
```diff
@@ -0,0 +1,73 @@
+      <message name="AUTOPILOT_VERSION"      period="11.1"/>
+      <message name="ROVER_STATUS"           period="1.2"/>
+      <message name="ROTORCRAFT_FP"          period="0.25"/>
+      <message name="STEERING_ROVER_DATA"    period="0.5"/>
```
maybe a new telemetry file is not needed and can be merged with default_rover.xml

gautierhattenberger commented on conf/tools/pprzgcs.xml:
```diff
@@ -1,2 +0,0 @@
-<program command="$pprzgcs" name="PprzGCS" />
```
please don't remove this file

gautierhattenberger commented on sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c:
```diff
@@ -0,0 +1,102 @@
+/* rover_steering_guidance.c */
```
please add paparazzi license header

gautierhattenberger commented on sw/airborne/firmwares/rover/guidance/rover_guidance_steering.c:
```diff
@@ -0,0 +1,102 @@
+    if (fabs(omega)>0.0) {
+      delta = -atanf(omega * DRIVE_SHAFT_DISTANCE / speed);
+    }
+    delta *= 180/M_PI;
```
if needed, their are `RadOfDeg` and `DegOfRad` macros

jesusBV20 replied:
Cool! I didn't know that

gautierhattenberger commented on sw/airborne/firmwares/rover/main.c:
```diff
@@ -0,0 +1,3 @@
+// Empty... rovers only use main_ap...
```
why do you need this file ?

jesusBV20 replied:
If I don't add this main.c then I get this error when trying to compile with NPS target:

make[1]: *** No rule to make target '/home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main.c', needed by '/home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main.o'.  Stop.
make[1]: *** Waiting for unfinished jobs....
CC /home/jesus/workspace/repos/paparazzi/var/aircrafts/ST_Rover/nps/firmwares/rover/main_ap.o
make[1]: Leaving directory '/home/jesus/workspace/repos/paparazzi/sw/airborne'
make: *** [Makefile.ac:213: nps.compile] Error 2

gautierhattenberger commented on sw/airborne/modules/guidance/gvf/gvf.c:
```diff
@@ -457,3 +476,259 @@ bool gvf_sin_wp_alpha(uint8_t wp, float alpha, float w, float off, float A)
+
+bool gvf_line_wp1_wp2(uint8_t wp1, uint8_t wp2)
+{
+  float x1 = waypoints[wp1].enu_f.x;
```
I recommend to use the `waypoint_get_x|y` function instead of accessing the element directly

noether commented:
Cool, thanks Gautier! 

We will revise the code as suggested.

noether commented:
@gautierhattenberger we will let you know when this PR is ready again. We will reorganize the modes to have manual and auto (we will remove for now the assisted one). We will add a speed controller that we already do not have, and finally, we will see what to do with the telemetry, if we need any extra messages or not.

noether commented:
@gautierhattenberger  can you check whether we have addressed your suggestions correctly? Maybe we have forgotten something. It has been a long time ago and we have rebased all the (new) commits that were done in parallel from master too.

noether commented:
There are two new main functionalities in this update:

- We have added a PID (+ feedforward) controller for the control of the ground speed.
- Three modes: 1. Manual, 2. Cruise (it sets as desired velocity the last one measured) but the course is still manual, 3. Auto (course and speed are both autonomous).

@jesusBV20 , let us know whether there are any other major changes.

gautierhattenberger commented:
There must be something wrong with your merge or rebase on master branch with more than one thousand files changed. Can you check that first ? I can hardly see the changes.

jesusBV20 commented:
I think that I did something wrong when I tried to rebase from our master. We don't know how to undo that, so I have created a clean branch and included every file changed.

noether commented:
Yep, we really do not how to do it (we are not deep expertise with git xD). 

Because the branch was created a time ago, in order to make sure it will compile with the current master (no subsystems, etc), we pulled (rebase) the current master to the branch. Is there a better way, or less messy, to do this procedure? (I would love to know it because I have a branch that I created 2 years ago xD that I will like to PR soon).

In parallel, @jesusBV20 has managed to create a new branch from the recent master, so there is no problem creating a new PR.

Let us know what would be the most convenient.

noether commented:
Oki, we were doing some research about rebase, and merge. We cannot revert the changes in this branch. We will create a new PR that you can track easily :P. I will close this PR then.
